### PR TITLE
modernise packaging, and better manage optional depencencies

### PR DESCRIPTION
**Pull request by liam-o-marsh**
note: this dropped the pinning on both python and the dependencies, please tell me if something doesn't work in a way the tests didn't catch

### CODE REVIEW

File: qstack/compound.py [hunk 1]
```diff
@@ -108,8 +108,18 @@ def gmol_to_mol(fin, basis="def2-svp"):
...
 
-    from cell2mol.tmcharge_common import Cell, atom, molecule, ligand, metal
-    from cell2mol.tmcharge_common import labels2formula
+    try:
+        from cell2mol.tmcharge_common import Cell, atom, molecule, ligand, metal
+        from cell2mol.tmcharge_common import labels2formula
+    except ImportError:
+            print("""
+
+ERROR: cannot import ĉell2mol. Have you installed qstack with the \gmol\" option?\n\n
```
Comment (briling) [inline qstack/compound.py:117]: *cell2mol

File: qstack/regression/regression.py [hunk 2]
```diff
@@ -2,7 +2,18 @@
...
+    from sklearn.model_selection import train_test_split
+except ImportError:
+    print("""
+
+ERROR: cannot import scikit-learn. Have you installed qstack with the \"regression\" option?\n\n
+(for instance, with `pip install qstack[regression] or `pip install qstack[all]``)
+
+""")
+    raise
+
```
Comment (briling) [inline qstack/regression/regression.py:16]: A bunch of files in qstack.regression use sklearn, do you think it's possible to complain not here (and not in each of them) but e.g. in qstack.regression.__init__?

**Review by briling — COMMENTED** [review]
Thanks a lot Liam!
Questions:
* I am not sure where the dependencies' versions are taken from -- automatically from  requirements.txt?
* Do you think it makes sense to make [all] the default option and add [minimal] or something that depends only on numpy?
* I think we should also make pyscf an optional dependency so one can e.g. install [qml] and use B2R2. However this would require more changes of imports I guess. Should it be a separeate PR?

**Comment by briling** [comment]
am I doing it wrong? 
```(qstack-test) 19:11:35-xe@dhcp-173-dist-b-094:~> pip install qstack[all] git+https://github.com/lcmd-epfl/Q-stack@repackaging-PR
Collecting git+https://github.com/lcmd-epfl/Q-stack@repackaging-PR
  Cloning https://github.com/lcmd-epfl/Q-stack (to revision repackaging-PR) to /tmp/pip-req-build-o5iewgtn
  Running command git clone --filter=blob:none --quiet https://github.com/lcmd-epfl/Q-stack /tmp/pip-req-build-o5iewgtn
  Running command git checkout -b repackaging-PR --track origin/repackaging-PR
  Switched to a new branch 'repackaging-PR'
  branch 'repackaging-PR' set up to track 'origin/repackaging-PR'.
  Resolved https://github.com/lcmd-epfl/Q-stack to commit c480da86df31ffd94da37d05e8c3cd6e28603c9c
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: numpy in /home/xe/SOFT/miniconda3/envs/qstack-test/lib/python3.9/site-packages (from qstack==0.0.1+v0.0.1.214.gc480da8) (1.22.3)
Requirement already satisfied: scipy in /home/xe/SOFT/miniconda3/envs/qstack-test/lib/python3.9/site-packages (from qstack==0.0.1+v0.0.1.214.gc480da8) (1.10.0)
Requirement already satisfied: pyscf in /home/xe/SOFT/miniconda3/envs/qstack-test/lib/python3.9/site-packages (from qstack==0.0.1+v0.0.1.214.gc480da8) (2.0.1)
Requirement already satisfied: tqdm in /home/xe/SOFT/miniconda3/envs/qstack-test/lib/python3.9/site-packages (from qstack==0.0.1+v0.0.1.214.gc480da8) (4.66.0)
Requirement already satisfied: h5py>=2.7 in /home/xe/SOFT/miniconda3/envs/qstack-test/lib/python3.9/site-packages (from pyscf->qstack==0.0.1+v0.0.1.214.gc480da8) (3.6.0)
Collecting equistore-core@ git+https://github.com/lab-cosmo/equistore.git@e5b9dc365369ba2584ea01e9d6a4d648008aaab8#subdirectory=python/equistore-core (from qstack==0.0.1+v0.0.1.214.gc480da8)
  Using cached equistore_core-0.1.0.dev260-py3-none-linux_x86_64.whl
Collecting sympy (from qstack==0.0.1+v0.0.1.214.gc480da8)
  Downloading sympy-1.12.1-py3-none-any.whl.metadata (12 kB)
INFO: pip is looking at multiple versions of qstack[equio,gmol,qml,regression,wigner] to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement cell2mol; extra == "gmol" (from qstack[equio,gmol,qml,regression,wigner]) (from versions: none)
ERROR: No matching distribution found for cell2mol; extra == "gmol"
```

**Comment by briling** [comment]
`pip install qstack[wigner] git+https://github.com/lcmd-epfl/Q-stack@repackaging-PR` works so something wrong with `[gmol]`?

**Comment by liam-o-marsh** [comment]
>     * I am not sure where the dependencies' versions are taken from -- automatically from  requirements.txt?
Nothing specific for now... I guess I can specify the versions from requirements.txt as minimal versions?

>     * Do you think it makes sense to make [all] the default option and add [minimal] or something that depends only on numpy?
not currently possible I think (https://github.com/pypa/setuptools/pull/1503), unless I were to change the build backend from setuptools to something else, but... I'm not sure it's even possible since we have a setup.py file

>     * I think we should also make pyscf an optional dependency so one can e.g. install [qml] and use B2R2. However this would require more changes of imports I guess. Should it be a separeate PR?
at this point the easier thing to do would be to split qml/ into a separate package called qstack-qml?

**Comment by briling** [comment]
> I guess I can specify the versions from requirements.txt as minimal versions?

I think for reproducibility it's better to fix the major and minor versions (e.g. one of our tests fails with pyscf 2.5 because they changed something)

> not currently possible I think

Okay! I thought I saw an example how to do it but maybe with an older configuration. Then we'll just need to be clear in the readme

> at this point the easier thing to do would be to split qml/ into a separate package called qstack-qml?

Makes sense! Is it possible with minimal refactoring?

**Comment by briling** [comment]
Also there's an upcoming patch concerning the regression module which will lead to some conflicts sorry

**Comment by liam-o-marsh** [comment]
> Makes sense! Is it possible with minimal refactoring?
Not really. As things are in the python ecosystem, there is no standard way to do this in a single project, so unless we switch the backend to Pantsbuild or whatever, we will have to move qml to a different repo altogether.

> Also there's an upcoming patch concerning the regression module which will lead to some conflicts sorry
not a problem, I'm almost used to dealing with this sort of stuff.

> I think for reproducibility it's better to fix the major and minor versions (e.g. one of our tests fails with pyscf 2.5 because they changed something)
is it a numerical result specifically?
in any case, yeah I have no issue putting upper bounds on packages, but fixing major.minor seems like overkill.
plus on anything we publish to zenodo we would have a requirements.txt file or something, right?

> Okay! I thought I saw an example how to do it but maybe with an older configuration. Then we'll just need to be clear in the readme
yeah I think this is possible in setup.py but honestly it's not an issue if we put that info there. better do that than make users' lives more complicated anyway

**Comment by briling** [comment]
> > Makes sense! Is it possible with minimal refactoring?
> 
> Not really. As things are in the python ecosystem, there is no standard way to do this in a single project, so unless we switch the backend to Pantsbuild or whatever, we will have to move qml to a different repo altogether.

I would like to avoid having a different repo. Can we do it in the way it's done in equistore/metatensor?

> > I think for reproducibility it's better to fix the major and minor versions (e.g. one of our tests fails with pyscf 2.5 because they changed something)
> 
> is it a numerical result specifically? in any case, yeah I have no issue putting upper bounds on packages, but fixing major.minor seems like overkill. plus on anything we publish to zenodo we would have a requirements.txt file or something, right?

Agree with adding upper bounds (can take the current ones I guess except for pyscf but we're planning to change the minimal version to 2.5 soon anyway)

**Comment by briling** [comment]
Well I just think that for "scientific" packages it makes sense to fix versions more strictly (in contrast to e.g. tqdm which just has to be compatible) but I can be wrong

**Comment by liam-o-marsh** [comment]
what did you say pyscf-v2.5.0  broke? I ran the tests both with the minimal versions from requirements.txt and the latest available version from all packages, and the tests give the same results, give or take numerical noise

**Comment by briling** [comment]
Everything passes for me too but earlier either `test_otpd.py` or `test_excited.py` used to break! @YAY-C do you remember the problem?

**Comment by briling** [comment]
Found: #31

**Comment by briling** [comment]
My check was wrong! Tried again and `test_otpd.py` failed

**Comment by liam-o-marsh** [comment]
...right, of course CI breaks down.

File: tests/test_otpd.py [hunk 3]
```diff
@@ -9,7 +10,7 @@ def test_hf_otpd():
     mol = compound.xyz_to_mol(path+'/data/H2O.xyz', 'def2svp', charge=0, spin=0)
     dm = fields.dm.get_converged_dm(mol, xc="pbe")
     otpd, grid = fields.hf_otpd.hf_otpd(mol, dm, return_all=True)
-    assert(abs(otpd @ grid.weights-20.190382555868)<1e-10)
```
Comment (liam-o-marsh) [inline tests/test_otpd.py:12]: I think I somehow made pyscf 2.5.0 compatible with qstack with this one line

**Comment by YAY-C** [comment]
> Everything passes for me too but earlier either `test_otpd.py` or `test_excited.py` used to break! @YAY-C do you remember the problem?

so for linux user's, I think there's no problem!
From my ubuntu machine at work and home I can install qstack from `pip install git+https://github.com/lcmd-epfl/Q-stack.git@master` and all the tests work fine!

`test_otpd.py` breaks if you use a newer version of pyscf (explicitly `pyscf==2.2.0` ,vs the required `pyscf==2.0.1`)
**but for now this is only relevant for MacOS users, because for some reason (C-compiler issues) you can not compile `pyscf==2.0.1`**

File: tests/test_otpd.py (hunk 3 above)
Comment (liam-o-marsh) [reply tests/test_otpd.py:12]: yeah look at those numbers:
```
print(otpd @ grid.weights, 20.190382555868)
20.19039044258982 20.190382555868
```

Comment (YAY-C) [reply tests/test_otpd.py:12]: mmh not sure it really solves compatibility, I think the threshold for accuracy is just lower with `np.allclose()` 
but I guess it makes a better test!
The assertion error in this case was not so far off anyway
```
 assert 7.886721846261935e-06 < 1e-10
```
_using `pyscf=2.2.0`_

**Comment by liam-o-marsh** [comment]
wouldn't be surprised if it was a change in the way the grid points are placed or something

**Comment by liam-o-marsh** [comment]
it's not the grid, and it's not libxc (it seems)
the 2.2.0 release notes, however, mention a change in the DIIS algorithm, so in case the DFT convergence tolerence isn't super tight...

**Comment by liam-o-marsh** [comment]
...can I have the ability to force-push here, please? So I can replace the final commit over and over to fix the CI

**Comment by briling** [comment]
I can't find how to do it. This branch isn't protected

**Comment by liam-o-marsh** [comment]
done!

**Comment by briling** [comment]
so cool thanks a lot!

**Comment by briling** [comment]
Can we still have max. version for pyscf? I've remembered that at some point they forced basis set order sorting wrt $\ell$ and it broke my function dealing with orca outputs

**Comment by liam-o-marsh** [comment]
ok!
I want to write a full test for this then. Do you know how to generate the right data to load?
